[depends: #472] Variable Sensitivity Domain

thk123 opened the pull request:
Beginning to integrate the "Variable Sensitivity Domain". This is a big change that was developed over a few months. I am prepared to split this into separate PRs to make it easier to review, but thought I'd start with everything so a) can see where it is going, b) if splitting is _not_ required, that saves me some work!

This PR depends on the work on goto-analyze #472 
This PR has unit tests so is dependent on #874 :white_check_mark: 

This PR includes the tests which Owen was working on, but these can go in as an independent PR #566 
:white_check_mark: 
Real Diff: https://github.com/diffblue/cbmc/pull/708/files/a8d47ce..1e23402 (excluding changes in the catch commit)

**Overview**

There is quite a lot of infrastructure that one needs to implement a non-relational domain
correctly.  For example, what you do about fields, setting things to top when you write to an unknown pointer, element-wise merging, etc.  The vast majority of that is independent of what per-variable abstraction you are using.  The variable sensitivity domain is an implementation of this infrastructure.

We have implemented constants and will implement intervals, in addition to the existing domains and goto-analyze will probably be using these in preference to the older ones.

There is some documentation in the header of `variable_sensitivity_domain.h` among other places. 

**Known limitations**

- Despite the flexibility of the architecture, not many options are implemented. Hopefully these will be coming soon (things like interval domain, squashed array etc)
- Functions sensitivity is not implemented (though if this is needed it is basically done so I can easily add to this PR once it is).
- Soundness on top pointers, out of bounds writing to arrays
- No support for unions

**Controversial elements**

- Double dispatch in the merge_state method :white_check_mark: 
- Each class requires hand written copy constructor :white_check_mark: 
- Abstraction hiding std::shared_ptr is leaky

**Splitting this up**

As mentioned above I could split this into the following PRs (with some work as there is some intermingling of changes):

Tests (#566) should go in first if going in separately. 
Core framework (just the base classes giving only the top/bottom abstraction for all types) and has all the tests turned off.

Then all of the following PRs would be independent of each other and would just enable a new abstract object:

1. Constant value
2. Constant array
3. Constant struct
4. Constant pointer

Before I do this, it probably still makes sense to review in this manner, i.e. review the following files first before moving on to the abstract objects:

- abstract_enviroment
- abstract_object
- abstract_value
- array_abstract_object
- pointer_abstract_object
- struct_abstract_object
- variable_sensitivity_domain
- varaible_sensitivity_object_factory
- goto_analyzer_parse_options
- static_analyzer
- static_show_domain
- static_simplifier

thk123 commented:
Before this gets merged, @martin-cs wanted to modify how the command line arguments are handled to mean we don't have to break backwards compatibility when more sensitivities are added.

thk123 commented on src/analyses/variable-sensitivity/constant_abstract_value.cpp:
```diff
@@ -0,0 +1,85 @@
+
+#include "constant_abstract_value.h"
+
+constant_abstract_valuet::constant_abstract_valuet(typet t):
```
This class is missing function header comments

thk123 replied:
Don.e

thk123 commented on src/analyses/variable-sensitivity/pointer_abstract_object.cpp:
```diff
@@ -0,0 +1,165 @@
+  }
+  else
+  {
+    return sharing_ptrt<pointer_abstract_objectt>(
```
This is wrong but I thought I'd fixed this somewhere...

thk123 replied:
Fixed in array_abstract_object but not elsewhere

thk123 commented on src/analyses/variable-sensitivity/struct_abstract_object.cpp:
```diff
@@ -0,0 +1,166 @@
+  else
+  {
+    return sharing_ptrt<struct_abstract_objectt>(
+      new struct_abstract_objectt(type(), false, true));
```
Again this is wrong - we should return top unless we are bottom

thk123 commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.cpp:
```diff
@@ -0,0 +1,345 @@
+    throw "exceptions not handled";
+
+  case OTHER:
+//    throw "other";
```
@martin-cs I had to comment this out early on as there are other instructions inside the CPROVER_Initailize I think - I suppose this is OK?

thk123 commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.cpp:
```diff
@@ -0,0 +1,345 @@
+  locationt to)
+{
+  #ifdef DEBUG
+  std::cout << "merging " << from->source_location.get_line() << "->"
```
This should use source location to be consistent with other debug output

thk123 commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.cpp:
```diff
@@ -0,0 +1,345 @@
+  return abstract_state.is_top();
+}
+
+
```
Too many empty lines

thk123 commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.cpp:
```diff
@@ -0,0 +1,345 @@
+bool variable_sensitivity_domaint::ai_simplify(
+  exprt &condition, const namespacet &ns, const bool lhs) const
+{
+  if (lhs)
```
To clarify the scope of this should be moved into a separate function `ai_simplify_lhs`

thk123 commented:
The following tests are failing and need fixing:

- [ ] constant_propagation_01
- [ ] constant_propagation_02

tautschnig commented:
Just reading the list of commits, there seems to be a lot of room for merging commits. Probably addressing the regression test failures is the first order of business, but then really a cleanup of those commits is due.

thk123 commented:
Fixed or disabled all the linting errors bar one - which was in a multi-line define so can't put the `// NOLINT` stuff. It is a bug in the linter as it is complaining about lack of braces around an if clause that has braces (but also has a `\` between the condition and the brace. 

@tautschnig Before I start squashing, would you like this as separate PRs as suggested above?

thk123 commented:
All the tests fail on Alpine Linux for an as of yet unknown reason.

martin-cs commented:
> +#include <iostream>
> +
> +bool abstract_environmentt::merge(const abstract_environmentt &env)
> +{
> +  // Use the sharing_map's "iterative over all differences" functionality
> +  // This should give a significant performance boost
> +  // We can strip down to just the things that are in both
> +
> +  // for each entry in the incoming environment we need to either add it
> +  // if it is new, or merge with the existing key if it is not present
> +
> +  std::stringstream merge_message;
> +
> +  ait<constant_propagator_domaint> ai;  // TODO - remove!
> 
> Remove! 

Yes!

> > +
> +  case RETURN:
> +    throw "return instructions are depreciated";
> +
> +  case START_THREAD:
> +  case END_THREAD:
> +  case ATOMIC_BEGIN:
> +  case ATOMIC_END:
> +    throw "threading not supported";
> +
> +  case THROW:
> +  case CATCH:
> +    throw "exceptions not handled";
> +
> +  case OTHER:
> +//    throw "other";
> 
> @martin-cs I had to comment this out early on as there are other instructions inside the CPROVER_Initailize I think - I suppose this is OK?

So there are a number of issues here...

1. Can we assume that return, throw/catch, function pointers,
vectors/complex, etc. have been removed?  I think this has to be case by
case.

1.A. return : cheap and exact to remove and later replace, makes the
analysis significantly simpler.  I think we assume that this has been
done.  Thus we should throw an exception if we find one and the
exception / code should document this.

1.B. throw/catch : This one is a bit more tricky as the removal process
would really benefit from being able to run abstract interpretation,
see:

https://github.com/diffblue/cbmc/pull/510

So I think, ideally, we should handle throw/catch.  We should probably
create an enhancement issue for this and make sure the code says "this
should be supported one day ... but isn't now".

1.C. function pointers : Same argument as 1.B.  the issues are already
raised:

https://github.com/diffblue/cbmc-toyota/issues/31   (workaround)
https://github.com/diffblue/cbmc-toyota/issues/30   (fix)

1.D. vectors/complex : I feel these are more like 1.A. so no action is
required (other than calling remove_vectors in goto-analyze which we
already do).

1.E. Concurrency ... should be handled in a different way.  We detect
concurrent programs and use the right ait<> variant.  Once we've done:

https://github.com/diffblue/cbmc-toyota/issues/85

we should be able to use this domain in that case but still start and
end thread should not need to be supported.



2. What should we do about unsupported instruction types?  Do we ignore
it so that we are robust to changes in other parts of the code base, or
do we throw an exception.  I'm inclined to say we ignore OTHER as it is
the 'escape hatch' for this kind of functionality but to throw an
exception for any other case.  But I'm open to suggestions for
alternatives.


Ideally we should probably "capture" this "design logic" in the code.


> +bool variable_sensitivity_domaint::merge(
> +  const variable_sensitivity_domaint &b,
> +  locationt from,
> +  locationt to)
> +{
> +  #ifdef DEBUG
> +  std::cout << "merging " << from->source_location.get_line() << "->"
> 
> This should use source location to be consistent with other debug output

Yes; consistency is output is good.


> > +   ns - the namespace
> +   lhs - is the expression on the left hand side
> +
> + Outputs: True if simplified the condition. False otherwise. condition
> +          will be updated with the simplified condition if it has worked
> +
> + Purpose: Use the information in the domain to simplify the expression
> +          with respect to the current location.  This may be able to
> +          reduce some values to constants.
> +
> +\*******************************************************************/
> +
> +bool variable_sensitivity_domaint::ai_simplify(
> +  exprt &condition, const namespacet &ns, const bool lhs) const
> +{
> +  if (lhs)
> 
> To clarify the scope of this should be moved into a separate function `ai_simplify_lhs`

Perhaps we should continue this as part of 

https://github.com/diffblue/cbmc-toyota/issues/108

martin-cs commented:
On Tue, 2017-03-28 at 02:28 -0700, Thomas Kiley wrote:
<snip>
> **Overview**
> 
> The variable sensitivity domain intends to replace all non-relational domains for goto-analyzer,

I might phrase this slightly differently.  There is quite a lot of
infrastructure that one needs to implement a non-relational domain
correctly.  For example, what you do about fields, setting things to top
when you write to an unknown pointer, element-wise merging, etc.  The
vast majority of that is independent of what per-variable abstraction
you are using.  The variable sensitivity domain is an implementation of
this infrastructure.

We have implemented constants and will implement intervals, in addition
to the existing domains and goto-analyze will probably be using these in
preference to the older ones but we didn't start out with the goal of
replacing them.

>  allowing flexibility in sensitivity at a very high grain of control.

(and thus a fair performance / accuracy trade-off)

> There is some documentation in the header of `variable_sensitivity_domain.h`. 

and elsewhere.

> **Known limitations**
> 
> - Despite the flexibility of the architecture, not many options are implemented. Hopefully these will be coming soon (things like interval domain, squashed array etc)
and should be able to be implemented with minor / non-intrusive patches.

> - Functions sensitivity is not implemented (though if this is needed it is basically done so I can easily add to this PR once it is).
Context sensitivity is something we will have to handle at the ait level
though.

> - Soundness on top pointers, out of bounds writing to arrays
(and more generally, what we do about undefined behaviour).

> **Controversial elements**
> 
> - Double dispatch in the merge_state method
> - Each class requires hand written copy constructor

See also:
https://github.com/diffblue/cbmc-toyota/issues/65

thk123 commented:
The failure on Alpine Linux doesn't seem to be strictly related to this PR, I have created a PR exhibiting the broken behaviour: #726

thk123 commented:
Today I will attempt to clean up the commit history, let me know @tautschnig if you want this split up, but I am assuming not.

martin-cs commented:
On Thu, 2017-03-30 at 02:33 -0700, Thomas Kiley wrote:
> Today I will attempt to clean up the commit history, let me know @tautschnig if you want this split up, but I am assuming not. 

Splitting makes sense for unrelated changes.  All of these are related
so a sequence is probably best.

thk123 commented:
Just as a progress update, I am getting there with the squashing, I am down to <100 reasonable commits which almost all compile, still to do is:

- [ ] Squash the struct abstract object related commits (these are the ones that have no-compiling steps)
- [ ] Extract all the hunks relating to Owens tests that are already in this PR: #566 so that it is easy to rebase once this is merged
- [ ] Have the tests turn themselves on at appropriate points in the commit history

Then it should be good for a proper review - I'm still happy to split into functional parts to speed up the review process.

thk123 commented:
Hopefully that is a sufficiently squashed commit history - I will now extract the changes that are part of the test PR

thk123 commented:
@martin-cs You are looking at the first commit - this way madness lies! This is addressed in: Cleaned up the variable sensitivity framework later down. No reasonable way to squash this into the first commit, too much changed in the meantime whose history would be bad to lose.

martin-cs commented:
On Mon, 2017-04-03 at 07:28 -0700, Thomas Kiley wrote:
> @martin-cs You are looking at the first commit - this way madness lies! This is addressed in: Cleaned up the variable sensitivity framework later down. No reasonable way to squash this into the first commit, too much changed in the meantime whose history would be bad to lose. 

It's worse than that; I'm responding to e-mail with minimal contextual
information.  Also we should get the ducks out of here.

thk123 commented:
I have now done all the rebasing - once #472 is merged I will rebase onto master. The real diff is this: https://github.com/diffblue/cbmc/pull/708/files/93b9278..6560460 (I've updated the comment at the top). This should be good for a proper review at this stage.

thk123 commented:
Rebased on current version of #472 (and dropped tests that have been merged into master). Updated real diff: https://github.com/diffblue/cbmc/pull/708/files/8ed388d..65fd967

thk123 commented:
@tautschnig What can I do to help get this reviewed? 

The linting errors are not real but also not disable-able due to being in a preprocessor macro. 

The failures on AppVeyor and on Alpine Linux due to #726. I can apply a fix to this by replacing asserts with CPROVER_asserts or I can start digging into #726 if that is what this is waiting for. 

I can attempt to break this up into smaller PRs if it is too large to review (alternatively, I can advise on an order to review the files in).

tautschnig commented:
A couple of thoughts:
1. Get #472 merged as both #472 and this one are fairly big steps.
2. If there are any bug fixes/smaller independent improvements in this PR, move them into a separate PR.
3. If within this PR, and on top of #472, any commits fix issues introduced by earlier commits in this branch then merge them.
Right now, it's a bit hard to see where to start, as neither the list of commits nor the full list of changes would provide just the view on this particular aspect. If you can post an updated link that has just the changes for variable-sensitivity that would be very helpful!
Thanks!

thk123 commented:
@tautschnig 

> 1. Get #472 merged

Yup though at the moment this is out of my/Dans hands

> 2. If there are any bug fixes/smaller independent improvements in this PR, move them into a separate PR.

Only bug fixes to this feature which seems nonsensical to split out.

> 3.  Merge fixing commits

I have to the best of my ability squashed the commits. 

As to where to start, [this is the only sensible smaller diff I can provide](https://github.com/diffblue/cbmc/pull/708/files/8ed388d7cd5e53598f167b0251f7362609466952..09765d01601c5ef52af3d93ece920619f9d5bd7b) which takes the feature up to point before including sensitivity for arrays, structs and pointers.

Perhaps I should sit down with someone in person (e.g. @peterschrammel) and talk through it? Ideally [Martin's documentation](https://github.com/diffblue/cbmc/pull/708/files/8ed388d7cd5e53598f167b0251f7362609466952..65fd967b646e41660513f646aa698a6deb90f03f#diff-f70d97a2bd3524c24857e91f9a97464cR7) should provide a guide to these classes:

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+{
+  assert(!is_bottom);
+  typedef std::function<abstract_object_pointert(const exprt &)> eval_handlert;
+  std::map<irep_idt, eval_handlert> handlers=
```
I do accept that the code below is correct, and maybe it's even beautiful. But it's probably quite a bit longer than a sequence of 5 `if`/`else if` statements, isn't it? The overhead of constructing the map might also outweigh the performance gains of doing log-N lookups instead of going through the cascade of ifs.

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+bool abstract_environmentt::assign(
+  const exprt &expr, const abstract_object_pointert value)
+{
+  exprt s = expr;
```
Throughout this function (and possibly the entire PR): no space around `=`, `==`, `!=`

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+    {
+      // Attempting to assign to something unreasonable
+      // Your goto-program is broken
+      throw "die_horribly";
```
This isn't a useful error message. Indeed an `assert` may be more appropriate.

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+    }
+  }
+
+  symbol_exprt symbol_expr=to_symbol_expr(s);
```
Do you genuinely need a copy, or would a reference do?

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+    // write_member which will attempt to update the abstract object for the
+    // relevant member. This modified abstract object is returned and this
+    // is inserted back into the map
+    std::map<irep_idt,stacion_functiont> handlers=
```
See my above comment for those handler maps.

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+  const typet type, bool top, bool bottom) const
+{
+  // TODO (tkiley): Here we should look at some config file
+  if(type.id()==ID_signedbv || type.id()==ID_bool || type.id()==ID_c_bool)
```
What do those types have in common, and why are other types not part of this disjunction? (Not all of them are bitvector types, and it isn't all bitvector types either.)

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+\*******************************************************************/
+
+abstract_object_pointert abstract_environmentt::abstract_object_factory(
+  const typet type, bool top, bool bottom) const
```
Why is `type` passed by value?

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+\*******************************************************************/
+
+abstract_object_pointert abstract_environmentt::abstract_object_factory(
+  const typet type, const constant_exprt e) const
```
Why are those arguments passed by value?

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+  const typet type, const constant_exprt e) const
+{
+  assert(type==e.type());
+  if(type.id()==ID_signedbv || type.id()==ID_bool || type.id()==ID_c_bool)
```
See above: unclear set of types.

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+  else
+  {
+    bool modified=false;
+    for(const auto &entry:env.map)
```
Nit picking: space around `:`

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+        abstract_object_pointert new_object=map[entry.first]->merge(
+          entry.second, object_modified);
+
+      modified|=object_modified;
```
Wrong indent

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+    {
+      if(env.map.find(entry.first)==env.map.end())
+      {
+        to_remove.push_back(entry.first);
```
I'd suggest erase-as-you-go instead of doing redundant lookups afterwards, unless this is infeasible for some reason.

thk123 replied:
This of course won't work as it is a map (d'oh) so have done it inline by incrementing the iterator (unless there is a nicer way I am unaware of @reuk?) 

```
const auto &end_iter=map.end();
for(auto iter=map.begin();iter!=end_iter; ++iter)
{
  if(env.map.find(iter->first)==env.map.cend())
  {
    modified=true;
    map.erase(iter++);
  }
}
```

tautschnig replied:
Isn't it `iter=map.erase(iter);` plus not incrementing `iter` in the loop header?

thk123 replied:
Yes of course, this is nonsense - serves me right for copying code (badly) without really considering it. Also reveals there is evidently at least one two few tests since this didn't fail.

thk123 replied:
Addressed in this commit: https://github.com/diffblue/cbmc/pull/708/commits/f69f28b39d43660e277a68f9ab8ec7f3df94d5ee

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,585 @@
+  out << "}\n";
+}
+
+abstract_object_pointert abstract_environmentt::eval_expression(
```
No function comment header

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.h:
```diff
@@ -0,0 +1,69 @@
+ Author: Thomas Kiley, thomas.kiley@diffblue.com
+
+\*******************************************************************/
+#ifndef CPROVER_ANALYSES_VARIABLE_SENSITIVITY_ABSTRACT_ENVIROMENT_H
```
Please add a blank line

thk123 replied:
Fixed.

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.h:
```diff
@@ -0,0 +1,69 @@
+
+  // This should be used as a default case / everything else has failed
+  // The string is so that I can easily find and diagnose cases where this occurs
+  virtual void havoc(const std::string &havoc_string);
```
This seems to have a single use case. I'm not sure that's a good code/value ratio.

tautschnig commented on src/analyses/variable-sensitivity/abstract_enviroment.h:
```diff
@@ -0,0 +1,69 @@
+
+  // For the less havocy pointer we might need this, which can be applied
+  // eagerly or lazily
+  //virtual void havoc (std::string s, typet t);
```
This is (obviously) never used.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+
+\*******************************************************************/
+
+#include <iostream>
```
iosfwd should do, I think

thk123 replied:
Unless I am mistaken, don't we need the full header in the cpp file?

tautschnig replied:
Yes, sorry, will likely be unavoidable.

tautschnig replied:
Yep, sorry.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+#include <analyses/variable-sensitivity/constant_abstract_value.h>
+
+#include <util/simplify_expr.h>
+#include <util/namespace.h>
```
Group includes by directory

thk123 replied:
Done.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+\*******************************************************************/
+
+abstract_objectt::abstract_objectt(const constant_exprt &expr):
+type(expr.type()), top(true), bottom(false)
```
Use leading space or don't - but do it consistently.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+type(expr.type()), top(true), bottom(false)
+{}
+
+abstract_objectt::~abstract_objectt()
```
Missing comment header

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+bool abstract_objectt::merge_state(
+  const abstract_object_pointert op1, const abstract_object_pointert op2)
+{
+  top=op1->top||op2->top;
```
Add space around `||`

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+  const abstract_object_pointert op, bool &out_any_modifications)
+{
+  assert(type==op->type);
+  abstract_object_pointert m=abstract_object_pointert(new abstract_objectt(*this));
```
Will be a candidate for `std::unique_ptr` - @reuk may be able to advise.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+    abstract_object_pointert lhs_abstract_object=environment.eval(op, ns);
+    const exprt &lhs_value=lhs_abstract_object->to_constant();
+
+    if(lhs_value.id()==ID_nil)
```
`lhs_value.is_nil()`

thk123 replied:
Fixed.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+      // One of the values is true so we can't really do anything more with
+      // this expression and should just return top for the result
+      return abstract_object_pointert(
+        new abstract_objectt(expr.type(), true, false));
```
How/why is this different from the `environment.abstract_object_factory(expr.type(), true, false);` ?

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+
+\*******************************************************************/
+
+bool abstract_objectt::is_bottom() const
```
Elsewhere there is `get_is_bottom` (and `get_is_top`) - consistent naming would be appreciated.

thk123 replied:
Yup addressed in this commit: e293073

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -0,0 +1,299 @@
+
+exprt abstract_objectt::to_constant() const
+{
+  return nil_exprt();
```
The comments above appear to suggest a rather different implementation.

thk123 replied:
Fixed.

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -0,0 +1,104 @@
+    return new current_typet(*this); \
+  } \
+
+#define MERGE(parent_typet) \
```
Why is this necessary - wouldn't a virtual `merge_state` do the trick?

tautschnig commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -0,0 +1,104 @@
+  virtual abstract_objectt* clone() const \
+  { \
+    typedef std::remove_const<std::remove_reference<decltype(*this)>::type \
+      >::type current_typet; \
```
Yes, this is valid C++11. Yet adding a simple `typedef some_classt selft;` to each of those classes and then doing `return new selft(*this);` might make for much simpler code. [But I'm yet to find out whether all this is really necessary at all.]

tautschnig commented on src/analyses/variable-sensitivity/abstract_value.h:
```diff
@@ -0,0 +1,26 @@
+#include <analyses/variable-sensitivity/abstract_object.h>
+
+
+class abstract_valuet:public abstract_objectt
```
What's the difference between an abstract value and and abstract object? I would have expected to have only one of those, with either name.

thk123 replied:
Conceptually this is the idea of a primitive to mirror the array, pointer union and struct abstract objects. Me and @martin-cs discussed this, the conclusion to keep it was based on some of the transform resolution code would end up in here.

tautschnig commented on src/analyses/variable-sensitivity/array_abstract_object.cpp:
```diff
@@ -0,0 +1,165 @@
+array_abstract_objectt::array_abstract_objectt(const typet &t):
+  abstract_objectt(t)
+{
+  assert(t.id()==ID_array);
```
Why not restrict the constructor to `const array_typet &`?

thk123 replied:
To keep the constructor consistent so can use in the abstract object factory: https://github.com/diffblue/cbmc/pull/708/files#diff-376bc05dbdc3369030faf77361786f1fR100

tautschnig commented on src/analyses/variable-sensitivity/array_abstract_object.cpp:
```diff
@@ -0,0 +1,165 @@
+  const typet &t, bool tp, bool bttm):
+    abstract_objectt(t, tp, bttm)
+{
+  assert(t.id()==ID_array);
```
See above.

tautschnig commented on src/analyses/variable-sensitivity/array_abstract_object.cpp:
```diff
@@ -0,0 +1,165 @@
+\*******************************************************************/
+
+abstract_object_pointert array_abstract_objectt::read_index(
+  const abstract_environmentt &env, const index_exprt &index) const
```
As the `index` parameter is never used here I don't quite know what this function is supposed to do, even in a more precise implementation. The same holds for `write_index`.

tautschnig commented on src/analyses/variable-sensitivity/constant_abstract_value.h:
```diff
@@ -0,0 +1,45 @@
+
+  virtual exprt to_constant (void) const;
+
+  virtual void output(
```
Why does this one have an `output` function, when others don't?

tautschnig commented on src/analyses/variable-sensitivity/pointer_abstract_object.cpp:
```diff
@@ -0,0 +1,160 @@
+pointer_abstract_objectt::pointer_abstract_objectt(const typet &t):
+  abstract_objectt(t)
+{
+  assert(t.id()==ID_pointer);
```
See comment regarding arrays above.

tautschnig commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.cpp:
```diff
@@ -0,0 +1,281 @@
+    break;
+
+  case RETURN:
+    throw "return instructions are depreciated";
```
Nit: typo "depreciated" -> "deprecated" (I'm not actually sure they are, though)

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.h:
```diff
@@ -43,6 +43,7 @@ class optionst;
   "(reachable-functions)" \
   "(intervals)(show-intervals)" \
   "(non-null)(show-non-null)" \
+  "(variable)" \
```
Somewhere there should be an update to a `--help` output.

tautschnig commented:
As I seem to have been commenting on some outdated version based on the link to a smaller diff: @thk123 would you mind scrutinising my recent comment even when marked "outdated", please?

thk123 commented on src/goto-analyzer/goto_analyzer_parse_options.h:
```diff
@@ -38,11 +130,19 @@ class optionst;
   "(intervals)(show-intervals)" \
-  "(non-null)(show-non-null)"
+  "(non-null)(show-non-null)" \
+  "(variable)" \
```
Somewhere there should be an update to a --help output.

thk123 commented:
Irritatingly, because they are "outdated", I can't reply inline, sorry. The "outdated" ones that I agree with I have moved to a new review on the latest code. I have left the ones that aren't outdated that I agree with and will do. Below are outdated ones I've not been able to reply to inline: 

> Use of map in place of chained ifs

My view was this made it easier to see which cases were handled. However I can swap it to `if`s if that is preferred. However, getting a third opinion, they agree with you since unused to seeing maps used this way. 

> Space around `=`

This one has been fixed, the linter has no errors. I found one other but I think this is otherwise not an issue any more. 

> This isn't a useful error

Yup - changed in later commit: https://github.com/diffblue/cbmc/pull/708/files#diff-7260d207dbd5603305c9641edbcd53b5R213

> Nit picking: space around :

Resolved

> This seems to have a single use case. I'm not sure that's a good code/value ratio.

I'm not sure I understand what action you want me to take on this?

> iosfwd should do, I think

Unless I am mistaken, don't we need the full header in the cpp file? 

> Add space around ||

Fixed

> How/why is this different from the `environment.abstract_object_factory(expr.type(), true, false);` ?

Fixed: https://github.com/diffblue/cbmc/pull/708/files#diff-487d940b8f9e1c792514613569de94d4R190

> Elsewhere there is get_is_bottom (and get_is_top) - consistent naming would be appreciated.`

Yup addressed in this commit: https://github.com/diffblue/cbmc/pull/708/commits/e29307337bf7ec632de059a7233d5353ce6a1215

> Why is this necessary - wouldn't a virtual merge_state do the trick?

I am deeply uncomfortable with this code as it is confusing without a long analysis. The reason for the define and virtual is to do double dispatch on the two types being merged. The logic being it finds the meet of the two classes, allowing for different sensitivities to combine. 

In practise we only support merging of the same type at type at the moment. An alternative that @martin-cs was opposed to was having some kind of separate data structure not tied to the class hierarchy which allowed looking up merges for two arbitrary abstract objects. 

> What is purpose of `read`/`write_index`

These provide the interface for sensitive abstract objects to read from or write to an array at an index. Perhaps looking at the constant_array_sensitivity_object implementations of these methods will make it clearer: https://github.com/diffblue/cbmc/pull/708/files#diff-5a47f0c5c6b007edfa773ce075ad1863R207

> Why does this one have an output function, when others don't

The base class implements the least sensitive output (TOP/BOTTOM) if a derived class provides greater sensitivity (for example, this class provides TOP/one value/BOTTOM) so it overrides output so that it can output this individual value. The array/pointer/struct classes only provide an interface, they do not increase the sensitivity of the object, and so do not need to override the base class.

tautschnig commented:
>> This seems to have a single use case. I'm not sure that's a good code/value ratio.

> I'm not sure I understand what action you want me to take on this?

[This was about havoc.] I'd suggest to remove `havoc`.

>> Why is this necessary - wouldn't a virtual merge_state do the trick?

> I am deeply uncomfortable with this code as it is confusing without a long analysis. The reason for the define and virtual is to do double dispatch on the two types being merged. The logic being it finds the meet of the two classes, allowing for different sensitivities to combine.

I'd suggest a chat with @reuk to devise a good solution.

thk123 commented on src/analyses/variable-sensitivity/variable_sensitivity_object_factory.cpp:
```diff
@@ -0,0 +1,153 @@
+
+  if(type.id()==ID_signedbv || type.id()==ID_unsignedbv ||
+    type.id()==ID_floatbv || type.id()==ID_fixedbv ||
+    type.id()==ID_c_bool || type.id()==ID_bool)
```
@tautschnig So I think this should be also primitive types, is there a definitive list or perhaps a function somewhere?

tautschnig replied:
You might want to do the inverse: list those types that are not handled? I'd go for `util/pointer_offset_size.cpp` as my first point of reference.

thk123 commented:
@tautschnig I am still a little confused about how you want me to remove havoc? It is used when, for example writing to a [top pointer](https://github.com/diffblue/cbmc/pull/708/files#diff-3346348fbcc9e93624981fc0e35bd5d0R156) so it is required.

tautschnig commented:
> @tautschnig I am still a little confused about how you want me to remove havoc? It is used when, for example writing to a top pointer so it is required.

Isn't that just a single use case? If so, I'd inline the implementation at this one point.

martin-cs commented:
On Wed, 2017-04-19 at 08:50 -0700, Michael Tautschnig wrote:
> > @tautschnig I am still a little confused about how you want me to remove havoc? It is used when, for example writing to a top pointer so it is required.
> 
> Isn't that just a single use case? If so, I'd inline the implementation at this one point.

The point of having a havoc function is to document / locate where
precision is being lost.  It may only be used in one place at the moment
but I don't expect that to last, esp. when we start implementing strict
modelling of undefined behaviour.

martin-cs commented:
On Wed, 2017-04-19 at 05:24 -0700, Michael Tautschnig wrote:
> >> Why is this necessary - wouldn't a virtual merge_state do the trick?
> 
> > I am deeply uncomfortable with this code as it is confusing without a long analysis. The reason for the define and virtual is to do double dispatch on the two types being merged. The logic being it finds the meet of the two classes, allowing for different sensitivities to combine.
> 
> I'd suggest a chat with @reuk to devise a good solution.

I've been through a large number of possible solutions.  I'm not saying
that you won't find a better one -- I hope you do but ... there are some
non-obvious challenges here.  The current solution is the sanest I found
so far.

martin-cs commented:
> Perhaps I should sit down with someone in person (e.g. @peterschrammel) and talk through it? Ideally [Martin's documentation](https://github.com/diffblue/cbmc/pull/708/files/8ed388d7cd5e53598f167b0251f7362609466952..65fd967b646e41660513f646aa698a6deb90f03f#diff-f70d97a2bd3524c24857e91f9a97464cR7) should provide a guide to these classes: 

I believe my documentation is now inline in the right places in the
code.  It may have even been updated...

tautschnig commented:
Having done a round of reviews and with @thk123 having implemented all the fixes we are now back to '#472 needs to be merged'.

thk123 commented:
@tautschnig @martin-cs @reuk @NathanJPhillips Thanks for those who've discussed the double dispatch part of the PR. 

The problems:

- Not immediately obvious what it is doing or why, or how to extend it
- lack of clarity where asymmetric merge methods should go (in fact I am not sure if they would even work without modifying the macro?)

Proposed solution:

- Remove merge, merge_state as they exist at the moment
- Add `protected virtual merge(abstract_object_pointert other) non-const`
- Each derived class does something like:
```C++
virtual constant_value::merge(abstract_object_pointert other) override
{
  if(other isa constant_value)
  {
    // do a constant value/constant value merge
  }
  else if(other isa interval) // Handle asymmetric merges here
  {
    // do a constant value/interval merge returning a constant value
  }
  else
  {
     super::merge(other);
  }
```

For dealing with the reflexive version, I think in general we actually need different code (if we want the result to be an interval when merging a constant and an interval we need different logic compared to what we need if we want a constant at the end. 

This nicely calls up to the parent. It does move the dynamic cast of the define, but I think makes it more clear. 

This would be accompanied by a static method on `abstract_objectt`:

```C++
abstract_object_pointert abstract_objectt::merge_objects(abstract_object_pointert op1, abstract_object_pointert op2)
{
  abstract_object_pointert result=op1->clone();
  result->merge(op2);
  return result;
}
```

This preserves immutability of the abstract objects and makes it obvious the return is the merge of the two params. The only thing I don't like is it isn't clear that the sensitivity of `op1` determines the sensitivity of the result.

This still requires the clone method, which I had hoped to get rid of 

@reuk kindly learnt me modern C++ design and I read the chapter on multiple dispatch. It outlines essentially three approaches, but the two run time approaches don't correctly respect inheritance so I don't think they are viable. The third approach involves some fairly heavy template meta programming machinery including explicitly declaring the inheritance hierarchy in a list which isn't ideal IMO.

I think if we do need to add this, we should at least wait until we need the double dispatch resolution so it is clearer what the problems will be. At the moment, we only have merge methods for matching types so the double dispatch is only uses insofar as to call up the chain if we have mixed sensitivities. 

If everyone is happy with this and can't spot any problems, I can either implement this as part of this PR or as a subsequent PR.

tautschnig commented:
I'm wondering whether it would be preferable to explicitly copy the `MERGE` (and possibly also `CLONE`) implementation into the respective classes, rather than relying on the preprocessor. Yes, there would be some duplication, but it might aid readability and be very very simple? No, it won't win us any beauty contest, sorry about that...

martin-cs commented:
On Thu, 2017-04-20 at 10:20 -0700, Michael Tautschnig wrote:
> I'm wondering whether it would be preferable to explicitly copy the `MERGE` (and possibly also `CLONE`) implementation into the respective classes, rather than relying on the preprocessor. Yes, there would be some duplication, but it might aid readability and be very very simple? No, it won't win us any beauty contest, sorry about that...

I had them as macros to reduce the amount of duplicate code and avoid
the risk of getting the dispatch wrong.  Plus, just to clarify the
design when I was first sketching it.  I'm still working through
@thk123's proposal which might remove the need for this, but, other than
that, if people don't want the macros, I'm fine with them being removed.

martin-cs commented:
On Thu, 2017-04-20 at 09:53 -0700, Thomas Kiley wrote:
> @tautschnig @martin-cs @reuk @NathanJPhillips Thanks for those who've discussed the double dispatch part of the PR. 
> 
> The problems:
> 
> - Not immediately obvious what it is doing or why, or how to extend it
> - lack of clarity where asymmetric merge methods should go (in fact I am not sure if they would even work without modifying the macro?)

I don't think they would either.  I think the only case I'd been
thinking of was ordered within the inheritance heirarchy, not the
unordered case.

> Proposed solution:
> 
> - Remove merge, merge_state as they exist at the moment

(FWIW merge_state was only introduced to split out the repetitive
boiler-plate and the actual merging of the state, plus it allows us to
get closer to the idea of having variable state as private).

> - Add `protected virtual merge(abstract_object_pointert other) non-const`
> - Each derived class does something like:
> ```C++
> virtual constant_value::merge(abstract_object_pointert other) override
> {
>   if(other isa constant_value)
>   {
>     // do a constant value/constant value merge
>   }
>   else if(other isa interval) // Handle asymmetric merges here
>   {
>     // do a constant value/interval merge returning a constant value
>   }
>   else
>   {
>      super::merge(other);
>   }
> ```

Yes; we will need something like this for transformers as well.


> For dealing with the reflexive version,

(Reflexive in class or in objects, for the object case (merging
something with itself) we could add a special case to just return the
object.)

>  I think in general we actually need different code (if we want the result to be an interval when merging a constant and an interval we need different logic compared to what we need if we want a constant at the end. 

(This is assuming that we have constant and interval as unordered /
independent children of value.)

> This nicely calls up to the parent.

Which is key to getting the "most general, common (parent) abstraction".

>  It does move the dynamic cast of the define, but I think makes it more clear. 
> 
> This would be accompanied by a static method on `abstract_objectt`:
> 
> ```C++
> abstract_object_pointert abstract_objectt::merge_objects(abstract_object_pointert op1, abstract_object_pointert op2)
> {

if (op1 == op2)  return op1;

>   abstract_object_pointert result=op1->clone();
>   result->merge(op2);
>   return result;
> }
> ```



> This preserves immutability of the abstract objects and makes it obvious the return is the merge of the two params.

Yes.

>  The only thing I don't like is it isn't clear that the sensitivity of `op1` determines the sensitivity of the result.

Yea... there is a problem here.  Imagine we have two abstractions,
parent and child.  Op1 is of type child and op2 is of type parent.  The
clone will give an object of type child but the merge will only merge
the state corresponding to the parent, leaving the rest of the child
state unmerged.

( There is also the awkwardness of requiring the merging of unordered
classes to be implemented in a symmetric way in two different places in
the code but without another layer of dispatch which swaps around the
ordering ... I'm not sure I see how to avoid this. )


> This still requires the clone method, which I had hoped to get rid of 

I'd suggest leaving it; it's just too useful.


> @reuk kindly learnt me modern C++ design and I read the chapter on multiple dispatch. It outlines essentially three approaches, but the two run time approaches don't correctly respect inheritance so I don't think they are viable. The third approach involves some fairly heavy template meta programming machinery including explicitly declaring the inheritance hierarchy in a list which isn't ideal IMO.

If this is Alexandrescu's book, then, yes, I had a look when I was first
working on this and came to a similar conclusions that his methods were
not really appropriate for what we wanted.  The key is that we really
want to merge / generate something of the most specific parent class in
almost all cases (the constant / interval one would be an exception);
that's what makes the current approach viable / more suitable than his
approach.

> I think if we do need to add this, we should at least wait until we need the double dispatch resolution so it is clearer what the problems will be. At the moment, we only have merge methods for matching types so the double dispatch is only uses insofar as to call up the chain if we have mixed sensitivities. 

This is the only use case I was thinking of.  Then again, as previously
discussed, it might be award to have intervals inherit from constant
(although, strictly speaking, correct), so we will have to live with the
unordered case soon enough.

> If everyone is happy with this and can't spot any problems, I can either implement this as part of this PR or as a subsequent PR. 

I like the design but am concerned about the problem given above.  If we
can fix that ... then we're in business.

thk123 commented:
> (Reflexive in class or in objects, for the object case (merging something with itself) we could add a special case to just return the object.)

Sorry used reflexive, meant reflective! 

> Op1 is of type child and op2 is of type parent.  The clone will give an object of type child but the merge will only merge the state corresponding to the parent, leaving the rest of the child state unmerged.

So the way I imagined this being handled, but didn't really explain in the pseudo-code, is that before the `super.merge`, the child class would essentially set its member variables to topish (for example, for the constant, set the `exprt value` to be nil). This still leaves param1 determining the type of the return. I don't really see a good alternative, because we don't want to choose the type statically, and in fact most of the time we do just want to mimic the existing type. I suppose we set it out explicitly:

```C++
abstract_objectt:create_from_merge(abstract_object op1, abstract_object op2, mutable_abstract_object result)
{
  result.merge(op1); // probably a NOOP 
  result.merge(op2);
}
```
Essentially assuming it is cloned from one of them, the merge should do nothing. The danger with this approach is you might naively pass it top as the third abstract object and you'd get top back. 

In any case, I'll get started, if anyone has any preference between the two interfaces or has a better suggestion let me know.

martin-cs commented:
A couple of extra thoughts on the topic:

1. Ideally merge should return a boolean of whether the object has been
changed or not, or, alternatively, make it trivial to find this (i.e.
returning the same pointer if and only if it is unchanged by the merge).
Having to explicitly compare the objects (say with ==) is overkill.

2. Transform / eval likely needs to follow a similar pattern of
dispatch, although it will hit the unordered class issue much faster,
for example pointer arithmetic.

thk123 commented:
I have made an improvement to the merge to not clone the abstract object unnecessarily as @martin-cs said this could be responsible for a big performance gain. I would appreciate some careful eyes over these changes as it involved using `enable_shared_from_this` which I've not used before @reuk @NathanJPhillips [diff](https://github.com/diffblue/cbmc/pull/708/files/c4daade66651873aac68796eb3c14af125195282..0ff6cc3090440aa1facc5eb59a971f758b850ac2)

I'm going to try and write some unit tests to verify that this optimisation is working correctly.

reuk commented on src/analyses/variable-sensitivity/constant_array_abstract_object.cpp:
```diff
@@ -127,34 +127,38 @@ Function: constant_array_abstract_objectt::constant_array_merge
     {
-      return false;
-    }
+      internal_sharing_ptrt<constant_array_abstract_objectt> result=
```
Looks like a good candidate for `auto`.

reuk commented on src/analyses/variable-sensitivity/constant_array_abstract_object.h:
```diff
@@ -50,7 +52,8 @@ class constant_array_abstract_objectt:public array_abstract_objectt
   CLONE
 
-  virtual bool merge(abstract_object_pointert other) override;
+  virtual abstract_object_pointert merge(
```
It's only possible to `override` `virtual` methods, therefore `virtual` is not required here.

reuk commented on src/analyses/variable-sensitivity/constant_array_abstract_object.h:
```diff
@@ -68,8 +71,8 @@ class constant_array_abstract_objectt:public array_abstract_objectt
-  bool constant_array_merge(
-    const constant_array_abstract_object_pointert other);
+  abstract_object_pointert constant_array_merge(
+    const constant_array_abstract_object_pointert other) const;
```
Declaring the parameter const in the interface doesn't do anything, and declaring it const in the implementation means that we can't move from the argument.

reuk commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.cpp:
```diff
@@ -161,32 +162,35 @@ Function: constant_pointer_abstract_objectt::merge_constant_pointers
-      value=nil_exprt();
-      assert(!is_bottom());
-      return !old->is_top();
+      internal_sharing_ptrt<constant_pointer_abstract_objectt> result=
```
Might look nicer with `auto`.

reuk commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.h:
```diff
@@ -48,12 +48,14 @@ class constant_pointer_abstract_objectt:public pointer_abstract_objectt
 
 protected:
-  virtual bool merge(abstract_object_pointert op1) override;
+  virtual abstract_object_pointert merge(
```
Unnecessary `virtual`.

reuk commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -97,7 +99,7 @@ class abstract_objectt
   abstract_object_pointert clone() const
   {
-    return mutable_clone();
+    return abstract_object_pointert(mutable_clone());
```
Is it necessary to convert this here? Is the return value of `mutable_clone` not automatically converted to `abstract_object_pointert`?

thk123 commented:
In keeping with the theme, I've written more unit tests and found more bugs. Basically the problem is when merging a struct/array abstract object, if the second object is top we are still trying to merge the maps, when in reality we should give up and return top. 

At the precise merge functions, I have been adding a clause that basically says when it is a simple merge:

```C++
if(is_top() || other->is_bottom())
```

Adding to this `other->is_top()` fixes this as it now calls the parent merge in this case. My question is, is this logic specific to the root abstract object, or does each level of sensitivity have some kind of "I can handle this merge, don't you worry about it, child class". 

In any case, I'll extract it out, but it isn't clear as one only have 1 (real) level of hierarchy, whether it should be virtual or not.

thk123 commented:
Rebased onto @martin-cs's goto-analyzer-5 branch, this is not doxygenated yet (that will come).

martin-cs commented:
Resolved by https://github.com/diffblue/cbmc/pull/5472
